pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go

```diff
@@ -574,3 +578,412 @@ func TestLoadBalancerWrongTierResourceDeletion(t *testing.T) {
 	_, err = gce.GetRegionAddress(lbName, region)
 	assert.True(t, isNotFound(err))
 }
+
+func TestEnsureExternalLoadBalancerFailsIfInvalidNetworkTier(t *testing.T) {
+	gce, err := fakeGCECloud()
+	require.NoError(t, err)
+
+	nodes, err := createAndInsertNodes(gce, []string{nodeName})
+	require.NoError(t, err)
+
+	// Enable the cloud.NetworkTiers feature
+	gce.AlphaFeatureGate.features[AlphaFeatureNetworkTiers] = true
+	wrongTier := "SupremeLuxury"
+	apiService.Annotations = map[string]string{NetworkTierAnnotationKey: wrongTier}
+
+	_, err = gce.ensureExternalLoadBalancer(clusterName, clusterID, apiService, nil, nodes)
+	require.Error(t, err)
+	assert.EqualError(t, err, fmt.Sprintf("unsupported network tier: %q", wrongTier))
+}
+
+func TestEnsureExternalLoadBalancerFailsWithNoNodes(t *testing.T) {
+	gce, err := fakeGCECloud()
+	require.NoError(t, err)
+
+	_, err = gce.ensureExternalLoadBalancer(clusterName, clusterID, apiService, nil, []*v1.Node{})
+	require.Error(t, err)
+	assert.EqualError(t, err, "Cannot EnsureLoadBalancer() with no hosts")
+}
+
+func TestForwardingRuleNeedsUpdate(t *testing.T) {
+	gce, err := fakeGCECloud()
+	require.NoError(t, err)
+	status, err := createExternalLoadBalancer(gce)
+	require.NotNil(t, status)
+	require.NoError(t, err)
+
+	lbName := cloudprovider.GetLoadBalancerName(apiService)
+	ipAddr := status.Ingress[0].IP
+
+	// When the loadBalancerIP does not equal the FwdRule IP address.
+	exists, needsUpdate, ipAddress, err := gce.forwardingRuleNeedsUpdate(lbName, region, "1.2.3.4", apiService.Spec.Ports)
+	assert.True(t, exists)
+	assert.True(t, needsUpdate)
+	assert.Equal(t, ipAddr, ipAddress)
+	assert.NoError(t, err)
+
+	// When loadBalancerPortRange returns an error.
+	exists, needsUpdate, ipAddress, err = gce.forwardingRuleNeedsUpdate(lbName, region, apiService.Spec.LoadBalancerIP, []v1.ServicePort{})
+	assert.True(t, exists)
+	assert.False(t, needsUpdate)
+	assert.Empty(t, ipAddress)
+	assert.NotNil(t, err)
+
+	// When portRange != the forwardingRule port range.
+	wrongPorts := []v1.ServicePort{apiService.Spec.Ports[0]}
+	wrongPorts[0].Port = wrongPorts[0].Port + 1
+	exists, needsUpdate, ipAddress, err = gce.forwardingRuleNeedsUpdate(lbName, region, apiService.Spec.LoadBalancerIP, wrongPorts)
+	assert.True(t, exists)
+	assert.True(t, needsUpdate)
+	assert.Equal(t, ipAddr, ipAddress)
+	assert.NoError(t, err)
+
+	// When the ports protocol does not equal the ForwardingRuel IP Protocol.
+	wrongPorts = []v1.ServicePort{apiService.Spec.Ports[0]}
+	wrongPorts[0].Protocol = v1.ProtocolUDP
+	exists, needsUpdate, ipAddress, err = gce.forwardingRuleNeedsUpdate(lbName, region, apiService.Spec.LoadBalancerIP, wrongPorts)
+	assert.True(t, exists)
+	assert.True(t, needsUpdate)
+	assert.Equal(t, ipAddr, ipAddress)
+	assert.NoError(t, err)
+
+	// Basic workflow.
+	exists, needsUpdate, ipAddress, err = gce.forwardingRuleNeedsUpdate(lbName, region, apiService.Spec.LoadBalancerIP, apiService.Spec.Ports)
```

Comment:
Many cases in this function seems possible to be tested with the same workflow, could we implement the test with a testcase struct? Something similar to below test func:
https://github.com/kubernetes/kubernetes/blob/f6eed81f211d929592ddb236535f306a277ff372/pkg/proxy/service_test.go#L97-L102

Reply:
Working on it.

Reply:
Done XD